### DISCUSSION TITLE
依赖究竟有哪些？以及如何设置后端代理？

### DISCUSSION THREAD

**Discussion by ivysrono** [SeaHOH/GotoX#12] [问答/咨询]
看糊涂了……

Release 3.3.4 里面有
> PySocks-1.6.6
> OpenSSL-16.2.0-cp35-win_amd64
> gevent-1.2.1-cp35-win_amd64
> dnslib-0.9.7
> 17MonIP-0.2.7

Master 里面仅有
> dnslib-0.9.7
> helpers_win32
> PySocks-1.6.6

ReadMe 里说：
> 必须组件：
>     gevent 1.1.2 及以上
>     pyOpenSSL 16.0.0 及以上
>     dnslib 0.8.3 及以上
>     PySocks 1.5.4 及以上

另外，我系统装了 Python3.6，直接把 .exe 复制过来就行？

**Comment by SeaHOH** [comment]
Master 一直在修改，所以 ReadMe 是指 Master，源码包含有部分依赖，但是不全；
Release 是发布当时的情况。

Windows 可以直接用发布的便携版本；
也可以下载源码使用，依赖就是 ReadMe 提到的；
要用 3.6 可以直接双击 start.py，但我没试过 3.6；
另外，我没有搞安装代码。

**Comment by ivysrono** [comment]
谢谢。这几天摸索下，从xxnet迁移过来，主要看重说明中的其他代理管理，这样可以免得再装浏览器扩展降低效率了。

**Comment by SeaHOH** [comment]
我只是简单测试过其它后端代理，欢迎报错。
另外虽然规则写了一大堆，其实也很简单，请仔细看一下 `Config.ini`，欢迎提问。

**Comment by ivysrono** [comment]
[proxy] 只能填一个？我想配置多个怎么办？

**Comment by SeaHOH** [comment]
主配置中的前置代理设置并未使用，配置了也不起作用。有时间会加上，优先级低。

后端代理可以在自动规则中随意配置，每条规则都可配置一个代理。

**Comment by ivysrono** [comment]
ActionFilter.ini
[12-Tor]
socks5://127.0.0.1:9150
www.tumblr.com

这样写似乎没用……
还是说只能是
11-proxy
12-proxy
然后定义不同的协议？

**Comment by SeaHOH** [comment]
是啊，通用的代理协议目前只有 **` HTTP`**、**`SOCKS4`** 和 **`SOCKS5`**，不然你浏览器也不支持不是。

**`proxy`** 就是我定义的代理动作命令关键字， 但 `Tor`  不是，使用说明里面已经写了。

以下为示例，实际上 `ActionFilter.ini` 文件已经写得很多了，也有默认配置的示例
```ini
[9-forward] #此小节为直接转发
.google.com = google_gws

[10-gae] #此小节为 GAE 代理
.greatfire.org
.google.com #上一个优先匹配，所以这个不会起作用

[11-proxy] #此小节为其它后端代理
#如果没有指定协议或指定了错误协议名称，会使用 HTTP 作为默认协议
.tumblr.com = socks5://127.0.0.1:9150
github.com = socks5://127.0.0.1:8080
```

**Comment by SeaHOH** [comment]
另外，以后问题和标题不太相干的，麻烦新开一个 issue，谢谢。

**Comment by ivysrono** [comment]
> 23:22:19 W 173.252.90.132 _create_ssl_connection 'www.tumblr.com' 返回 timeout('timed out',)，重试
> 23:22:19 W 173.252.90.132 create_ssl_connection 'https://www.tumblr.com/login?redirect_to=%2Fdashboard' 失败：timeout('timed out',)
> 23:22:19 W do_DIRECT "GET https://www.tumblr.com/login?redirect_to=%2Fdashboard" 失败，尝试使用 "GAE" 规则。

不管我换什么代理，http还是socks5，都是如上报错。

```
[9-proxy]
www.tumblr.com=socks5://127.0.0.1:9150
```

**Comment by SeaHOH** [comment]
没有正确设置。等号前面要有空格，后面无所谓。
`ActionFilter.ini` 第二句说的就是这个，麻烦先大略浏览一下。

**Comment by ivysrono** [comment]
好坑的错误……
然后Firefox许久提示安全链接失败，但我用 meow 去调用 Psiphon 和 Tor 都是可以的。
终端无错误信息输出。
![snipaste_20170319_001445](https://cloud.githubusercontent.com/assets/2649544/24073837/150ee460-0c39-11e7-94e8-2709229324b2.png)
![snipaste_20170319_001943](https://cloud.githubusercontent.com/assets/2649544/24073887/c73dd402-0c39-11e7-81b6-7d8696cc8bbe.png)

**Comment by SeaHOH** [comment]
有一条导入不知道什么时候误删了，现已修复。你可以下载主分支重试。

**Comment by ivysrono** [comment]
谢谢，终于成功了。
还有几个问题：

1. 17MonIP-0.2.7.egg 是用于更新其IP库？
2. 如果用 Python3.6，直接通过 start.py 启动，有无办法调用托盘功能？
3. 能否提供 ContentFilter.user.ini ?
4. ActionFilter.ini 中，[5-gae]、 [8-gae] 有没有区别？为什么跳过了4？

**Comment by SeaHOH** [comment]
1、现在不用地理库了，改为直连 IP 库，可以运行 **`launcher\buildipdb.py`** 更新，Windows 还可以通过托盘更新。
2、为了方便调用，托盘设计成通过便携版调用，要用安装版需改路径，不通用。
3、由于某个规则设计还没有好方案，所以暂时搁浅。
4、数字只是用于匹配时排序，与实际顺序无关，2 也可以放在 1 前面。[5-gae] 和 [8-gae] 的区别就是匹配优先级不同。这个文件里面有写啊。

**Comment by SeaHOH** [comment]
想了一下也可以用安装版启动托盘，等会儿改一下。

**Comment by ivysrono** [comment]
我去python官网下了便携版3.6解压放入 GotoX\python，但无法通过GotoX.vbs生成托盘。

**Comment by SeaHOH** [comment]
是的，因为还需要一些改动。还是等我修改成可以用安装版吧。

**Comment by SeaHOH** [comment]
已经改好，可以从安装版启动了。直接运行 **`launcher/win32.py`** 就行。

**Comment by ivysrono** [comment]
多谢，成功了。还实现了托盘上右键直接显示窗口的功能，这是我本来想提 issue 的。
额，还是有两个小问题：

1. 还是不能用便携版 Python3.6；
2. 通过 win32.py 启动的托盘是 32 位的，而通过 GotoX.vbs 启动的是 64 位的。

还是希望使用便携版 Python3.6，不希望其它各种库污染配置，或者是库升级导致功能失效。

**Comment by SeaHOH** [comment]
1、上面说了，官方便携版需要改动才可用于 GotoX。
2、你安装的什么版本，直接运行 win32.py 就是什么版本。

关于发布 3.6 便携版的问题，这个版本才出来，我还想再等一段时间。
